op-monitorism/transaction_monitor/monitor.go

```diff
@@ -0,0 +1,411 @@
+	}
+	rootClaim, ok := rootClaimResult[0].([32]byte)
+	if !ok {
+		return false, fmt.Errorf("invalid root claim returned")
+	}
+
+	var extraDataResult []interface{}
+	if err := game.Call(&bind.CallOpts{Context: ctx}, &extraDataResult, "extraData"); err != nil {
+		return false, fmt.Errorf("failed to get extra data: %w", err)
+	}
+	if len(extraDataResult) != 1 {
+		return false, fmt.Errorf("unexpected number of return values for extraData")
+	}
+	extraData, ok := extraDataResult[0].([]byte)
+	if !ok {
+		return false, fmt.Errorf("invalid extra data returned")
+	}
+
+	// Verify with factory
+	var factoryResult []interface{}
+	if err := factory.Call(&bind.CallOpts{Context: ctx}, &factoryResult, "games", gameType, rootClaim, extraData); err != nil {
+		return false, fmt.Errorf("failed to verify game with factory: %w", err)
+	}
+	if len(factoryResult) != 2 {
+		return false, fmt.Errorf("unexpected number of return values from factory")
+	}
+
+	proxy, ok := factoryResult[0].(common.Address)
+	if !ok {
+		return false, fmt.Errorf("invalid proxy address returned")
+	}
+
+	timestamp, ok := factoryResult[1].(uint64)
+	if !ok {
+		return false, fmt.Errorf("invalid timestamp returned")
+	}
+
+	isValid := proxy == gameAddr && timestamp > 0
+
+	// Cache the result
+	v.mu.Lock()
+	v.cache[gameAddr] = isValid
+	v.mu.Unlock()
+
+	return isValid, nil
+}
+
+func (m *Monitor) Run(ctx context.Context) {
+	// Initialize starting block
+	currentBlock := m.startBlock
+	if currentBlock == 0 {
+		var err error
+		currentBlock, err = m.client.BlockNumber(ctx)
+		if err != nil {
+			m.log.Error("failed to get initial block number", "err", err)
+			return
+		}
+		m.log.Info("starting from latest block", "number", currentBlock)
+	}
+
+	ticker := time.NewTicker(time.Second)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			m.log.Info("monitor stopping")
+			return
+		case <-ticker.C:
+			// Get latest block number
+			latestBlock, err := m.client.BlockNumber(ctx)
+			if err != nil {
+				m.log.Error("failed to get latest block number", "err", err)
+				continue
+			}
+
+			// Process blocks until we catch up to head
+			for currentBlock <= latestBlock {
+				if err := m.processBlock(ctx, currentBlock); err != nil {
+					m.log.Error("failed to process block", "number", currentBlock, "err", err)
+					// Only increment error counter for unexpected errors
+					if !errors.Is(err, ethereum.NotFound) {
+						m.unexpectedRpcErrors.WithLabelValues("monitor", "getBlock").Inc()
+					}
+					break
+				}
+				currentBlock++
+			}
+		}
+	}
+}
+
+func (m *Monitor) processBlock(ctx context.Context, blockNum uint64) error {
+	block, err := m.client.BlockByNumber(ctx, big.NewInt(int64(blockNum)))
+	if err != nil {
+		return fmt.Errorf("failed to get block: %w", err)
+	}
+
+	for _, tx := range block.Transactions() {
+		if tx.To() == nil {
+			continue
+		}
+
+		if config, exists := m.watchConfigs[*tx.To()]; exists {
```

Comment:
Oh this is backwards, we care about transactions going out of the account, not transactions coming into the account